Content.Server/Fluids/EntitySystems/AbsorbentSystem.cs

```diff
@@ -117,30 +117,43 @@ private bool TryTransferAbsorber(EntityUid user, EntityUid used, EntityUid targe
         if (!_solutionSystem.TryGetRefillableSolution(target, out var refillableSolution, refillable: refillable))
             return false;
 
-        if (refillableSolution.Volume <= 0)
+        if (refillableSolution.Volume < 0)
         {
             var msg = Loc.GetString("mopping-system-target-container-empty", ("target", target));
             _popups.PopupEntity(msg, user, user);
             return false;
         }
 
+        var transferAmount = component.PickupAmount / PuddleSystem.EvaporationReagentRatio -
+            absorberSoln.GetReagentQuantity(PuddleSystem.EvaporationReagent);
+
+        transferAmount = transferAmount > 0 ? transferAmount : 0;
+
         // Remove the non-water reagents.
         // Remove water on target
         // Then do the transfer.
-        var nonWater = absorberSoln.SplitSolutionWithout(component.PickupAmount, PuddleSystem.EvaporationReagent);
+        var water = refillableSolution.RemoveReagent(PuddleSystem.EvaporationReagent, transferAmount);
+
+        // Reads from refillableSolution.AvailableVolume have to
+        // be done after possibly removing some water
+        if (refillableSolution.AvailableVolume == FixedPoint2.Zero){
+            _popups.PopupEntity(Loc.GetString("mopping-system-full", ("used", target)), user, user);
+            // Had it removed any it wouldn't be full
+            DebugTools.Assert(water == FixedPoint2.Zero);
+            return false;
+        }
```

Comment:
Previously you could not unload dirt volume into a bucket with no water so a check like that if was not needed.
Now it might issue a full message if the container is full and either there weren't any water in the container to wet the mop, or if the mop already had enough water. Otherwise the container would not be full at this point of code execution.